plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -175,7 +187,17 @@ public String getProjectId()
 
     public Iterable<Dataset> listDatasets(String projectId)
     {
-        return bigQuery.listDatasets(projectId).iterateAll();
+        try {
+            return remoteDatasetCache.get(projectId);
+        }
+        catch (ExecutionException e) {
+            throw new TrinoException(BIGQUERY_LISTING_DATASET_ERROR, "Failed to retrieve datasets from BigQuery", e);
+        }
+    }
+
+    private List<Dataset> listDatasetsFromBigQuery(String projectId)
+    {
+        return Streams.stream(bigQuery.listDatasets(projectId).iterateAll()).collect(toImmutableList());
```

Comment:
nit: stream formatting

```suggestion
        return stream(bigQuery.listDatasets(projectId).iterateAll())
            .collect(toImmutableList());
```